### ISSUE TITLE
thunk/dispatch type is not getting inferred correctly together with redux toolkit

### ISSUE THREAD

**Issue by AlwaysBlind** [reduxjs/redux-toolkit#1048]
## Context: 
When using redux-thunk middleware together with redux-toolkit,
the thunk/dispatch types are not properly inferred.

I tried to follow the instructions given in [redux-toolkit.js](https://redux-toolkit.js.org/usage/usage-with-typescript#getting-the-dispatch-type)
to include the thunk-middleware and then dispatch a thunk. 

I created an example where this fails here: [codepen](https://codesandbox.io/s/relaxed-johnson-u8vu0?file=/src/index.ts).

## Expected result:
Should be able to dispatch a thunk without typescript error.

## What happened
Typescript could not infer the correct type for the dispatch/thunk and claims:

```No overload matches this call.
  Overload 1 of 2, '(action: AnyAction): AnyAction', gave the following error.
    Argument of type '(dispatch: AppDispatch, getState: () => State, extraArgs: string) => Promise<void>' is not assignable to parameter of type 'AnyAction'.
      Property 'type' is missing in type '(dispatch: AppDispatch, getState: () => State, extraArgs: string) => Promise<void>' but required in type 'AnyAction'.
  Overload 2 of 2, '(asyncAction: ThunkAction<Promise<void>, {}, string, AnyAction>): Promise<void>', gave the following error.
    Argument of type '(dispatch: AppDispatch, getState: () => State, extraArgs: string) => Promise<void>' is not assignable to parameter of type 'ThunkAction<Promise<void>, {}, string, AnyAction>'.
      Types of parameters 'getState' and 'getState' are incompatible.
        Property 'isLoading' is missing in type '{}' but required in type 'State'.ts(2769)
index.d.ts(21, 3): 'type' is declared here.
index.ts(4, 23): 'isLoading' is declared here.
```

## Relevant packages:
 * "@reduxjs/toolkit": "1.5.1"
 * "redux-thunk": "2.3.0"

## Steps to reproduce:

Same codepen as above: [codepen](https://codesandbox.io/s/relaxed-johnson-u8vu0?file=/src/index.ts)

## Note:
Here is a similar issue without redux-toolkit: [issue](https://github.com/reduxjs/redux-thunk/issues/288#issuecomment-744078267)

**Comment by markerikson** [comment]
You shouldn't need to reference the thunk middleware directly, and that is in fact likely the issue here.  See https://redux-toolkit.js.org/api/getDefaultMiddleware#customizing-the-included-middleware .

Try changing it to:

```ts
export const store = configureStore({
  reducer: userSlice.reducer,
  middleware: getDefaultMiddleware => getDefaultMiddleware({
    thunk: {
      extraArgument: ""
    }
  })
});
```

**Comment by AlwaysBlind** [comment]
This indeed solved it for me. Thank you!

**Comment by bjbrewster** [comment]
Thanks. You helped me fix a similar bug, but in my case I was exporting a function to create the store that can make unit tests easier.

```
import { configureStore, Store } from '@reduxjs/toolkit'

import rootReducer from './reducers'
import middlewares from './middleware'

export function configureAppStore(preloadedState: any): Store {
  return configureStore({
    reducer: rootReducer,
    middleware: getDefaultMiddleware => getDefaultMiddleware().concat(...middlewares),
    preloadedState,
  });
}

export type AppState = ReturnType<typeof rootReducer>
export type AppDispatch = ReturnType<typeof configureStore>['dispatch']
```

My `AppDispatch` type was always `Dispatch<Action<any>>` and not `ThunkAction<...> & Dispatch<...>` because `typeof configureStore` did not have the right template params by itself.

The solution was to change AppDispatch to `ReturnType<typeof configureAppStore>['dispatch']`  and remove the return type of configureAppStore (`: Store`) so it was correctly inferred.

**Comment by markerikson** [comment]
@bjbrewster I think the real issue there is that you're hard-setting the return type to be `Store`. Try removing that and let TS infer the return type of `configureAppStore` instead.

**Comment by bjbrewster** [comment]
Yep, I thought I had described that at the end of my comment.  Sorry
for the confusion.

**Comment by markerikson** [comment]
@bjbrewster you did - I read too quickly, wrote a response, then realized you'd already said that at the end of your comment :)
